cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/GoalBasedOperationRunnable.java

```diff
@@ -0,0 +1,172 @@
+ */
+
+package com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable;
+
+import com.linkedin.cruisecontrol.exception.NotEnoughValidWindowsException;
+import com.linkedin.kafka.cruisecontrol.KafkaCruiseControl;
+import com.linkedin.kafka.cruisecontrol.analyzer.OptimizerResult;
+import com.linkedin.kafka.cruisecontrol.analyzer.goals.Goal;
+import com.linkedin.kafka.cruisecontrol.async.progress.OperationProgress;
+import com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig;
+import com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException;
+import com.linkedin.kafka.cruisecontrol.monitor.ModelCompletenessRequirements;
+import com.linkedin.kafka.cruisecontrol.servlet.parameters.GoalBasedOptimizationParameters;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Pattern;
+
+import static com.linkedin.kafka.cruisecontrol.KafkaCruiseControlUtils.goalsByPriority;
+import static com.linkedin.kafka.cruisecontrol.KafkaCruiseControlUtils.sanityCheckGoals;
+import static com.linkedin.kafka.cruisecontrol.KafkaCruiseControlUtils.sanityCheckLoadMonitorReadiness;
+import static com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RunnableUtils.SELF_HEALING_DRYRUN;
+import static com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RunnableUtils.SELF_HEALING_STOP_ONGOING_EXECUTION;
+import static com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RunnableUtils.SELF_HEALING_MODEL_COMPLETENESS_REQUIREMENTS;
+import static com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RunnableUtils.SELF_HEALING_SKIP_HARD_GOAL_CHECK;
+import static com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RunnableUtils.SELF_HEALING_EXCLUDED_TOPICS;
+import static com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RunnableUtils.maybeStopOngoingExecutionToModifyAndWait;
+
+
+/**
+ * An abstract class to extracting the common logic of goal based operation runnables.
+ */
+public abstract class GoalBasedOperationRunnable extends OperationRunnable {
+  protected final List<String> _goals;
+  protected final ModelCompletenessRequirements _modelCompletenessRequirements;
+  protected final boolean _dryRun;
+  protected final boolean _stopOngoingExecution;
+  protected final boolean _skipHardGoalCheck;
+  protected final Pattern _excludedTopics;
+  protected final boolean _allowCapacityEstimation;
+  protected final boolean _excludeRecentlyDemotedBrokers;
+  protected final boolean _excludeRecentlyRemovedBrokers;
+  protected OperationProgress _operationProgress;
+  // Combined completeness requirements to be used after initialization.
+  protected ModelCompletenessRequirements _combinedCompletenessRequirements;
+  protected List<Goal> _goalsByPriority;
+
+  public GoalBasedOperationRunnable(KafkaCruiseControl kafkaCruiseControl,
```

Comment:
why `DemoteBrokerRunnable` does not extends from this new abstraction?

Reply:
This is preserving the abstraction that `DemoteBrokerParameters` does not extend from `GoalBasedOptimizationParameters`. The key reason is that the `demote` operation relies on a single hard-coded goal rather than using a variable goals for execution.